### PR TITLE
Close chat tabs with middle mouse button

### LINKED ISSUE #3418: Close chat tabs with middle mouse button
Similar to the tabs in Google Chrome, I'd like to request adding middle-click to close a chat tab.

### PR DESCRIPTION
**Pull request by LastExceed**
Clicks the close button when clicking a chat tab with mouseM.

- Closes #3418 

---

Clicking on a chat tab with the middle mouse button now closes the tab, just like in web browsers. -> [video](https://streamable.com/ivmbm)

### CODE REVIEW

**Review by peppy — DISMISSED** [review]

File: osu.Game/Overlays/Chat/ChatTabControl.cs [hunk 1]
```diff
@@ -143,6 +144,13 @@ private void fadeInactive()
                 textBold.FadeOut(transition_length, Easing.OutQuint);
             }
 
+            protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
+            {
+                if (args.Button == MouseButton.Middle)
+                    closeButton.Action();
+                return base.OnMouseUp(state, args);
```
[the remark this answers is not in the document]
Comment (LastExceed) [reply osu.Game/Overlays/Chat/ChatTabControl.cs]: done

**Comment by morguldir** [comment]
> I'd hope we can use `OnClick` instead of `OnReleased` in the future. @ekrctb do you recall a reason we didn't turn clicks on for `MinorMousebutton` managers?

If you use OnReleased you should be able to drag away to avoid closing it  if you clicked by accident, this is what web browsers do as well.

**Comment by peppy** [comment]
Yes, `OnClick` works in the same way.

**Comment by ekrctb** [comment]
Click cannot be enabled for now because `OnClick` doesn't have an argument to know which button is this click. Note that examining `InputState` to find the button doesn't work because there are cases when multiple buttons are pressed while a button is clicked.

**Comment by LastExceed** [comment]
plugging ppy/osu-framework#1594 for reference

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Overlays/Chat/ChatTabControl.cs [hunk 2]
```diff
@@ -143,6 +144,13 @@ private void fadeInactive()
                 textBold.FadeOut(transition_length, Easing.OutQuint);
             }
 
+            protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
+            {
+                if (args.Button == MouseButton.Middle)
+                    closeButton.Action();
+                return true;
```
Comment (ekrctb) [inline osu.Game/Overlays/Chat/ChatTabControl.cs]: should return true only when button is middle.

Comment (LastExceed) [reply osu.Game/Overlays/Chat/ChatTabControl.cs]: Oh right. Gonna fix that

Comment (peppy) [reply osu.Game/Overlays/Chat/ChatTabControl.cs]: I was okay with it always returning `true`, because channel tabs should never let input pass through.

Comment (ekrctb) [reply osu.Game/Overlays/Chat/ChatTabControl.cs]: But here is `OnMouseUp`. Maybe should block `OnMouseDown` additionally in that case.

Comment (smoogipoo) [reply osu.Game/Overlays/Chat/ChatTabControl.cs]: No big issue for me either way.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**